libraries/AP_Scripting/examples/DroneCAN_test.lua

```diff
@@ -0,0 +1,261 @@
+    local exponent = (v16 >> 10) & 0x1F
+    local fraction = v16 & 0x3FF
+
+    local value
+    if exponent == 0 then
+        if fraction == 0 then
+            -- Zero (positive or negative)
+            value = 0.0
+        else
+            -- Subnormal numbers (exponent = -14, no implicit leading 1)
+            value = (fraction / 1024.0) * 2.0^-14
+        end
+    elseif exponent == 0x1F then
+        if fraction == 0 then
+            -- Infinity (positive or negative)
+            value = math.huge
+        else
+            -- NaN (Not a Number)
+            value = 0/0
+        end
+    else
+        -- Normalized numbers: implicit 1 before the fraction and exponent bias of 15.
+        value = (1 + fraction / 1024.0) * 2.0^(exponent - 15)
+    end
+
+    -- Apply the sign bit
+    if sign == 1 then
+        value = -value
+    end
+
+    return value
+end
+
+--[[
+    check for incoming airspeed broadcast messages
+--]]
+local function check_airspeed()
+    local payload, nodeid = airspeed_handle:check_message()
+    if not payload then
+        return
+    end
+    local flags, static_pressure, differential_pressure, static_pressure_sensor_temperature,
+        differential_pressure_sensor_temperature, static_air_temperature, pitot_temperature = string.unpack("BffHHHH", payload)
+    if flags then
+        local temp_C = unpackFloat16(static_air_temperature) - 273.15;
+        gcs:send_text(MAV_SEVERITY.INFO, string.format("Rawairdata(%u): %f %.2fC",
+                                                       nodeid, differential_pressure, temp_C))
+        logger.write("ASPL", 'SP,DP,ST,DT,SAT,PIT', 'ffffff',
+                     static_pressure, differential_pressure, static_pressure_sensor_temperature, differential_pressure_sensor_temperature,
+                     static_air_temperature, pitot_temperature)
+    end
+end
+
+--[[
+    parse a parameter GetSet NumericValue
+--]]
+local function parse_param_NumericValue(payload, byte_offset)
+    local vtype = string.unpack("B", payload, byte_offset[1])
+    byte_offset[1] = byte_offset[1] + 1
+    if vtype == 0 then
+        return nil
+    elseif vtype == 1 then
+        -- integer (treat as 32 bit for now, actually 64 bit)
+        local ret = string.unpack("i", payload, byte_offset[1])
+        byte_offset[1] = byte_offset[1] + 8
+        return ret
+    elseif vtype == 2 then
+        -- float32
+        local ret = string.unpack("f", payload, byte_offset[1])
+        byte_offset[1] = byte_offset[1] + 4
+        return ret
+    else
+        return nil
+    end
+end
+
+--[[
+    parse a parameter GetSet Value
+--]]
+local function parse_param_Value(payload, byte_offset)
+    local vtype = string.unpack("B", payload, byte_offset[1])
+    byte_offset[1] = byte_offset[1] + 1
+    if vtype == 0 then
+        return nil
+    elseif vtype == 1 then
+        -- integer (treat as 32 bit for now, actually 64 bit)
+        local ret = string.unpack("i", payload, byte_offset[1])
```

Comment:
You should be able to unpack it to two ints and the reassemble as a uint64. I think the only thing we would be missing would be the logging of a int64 by casting from a uint64.

Also unpacking as 32bit will mean negative numbers don't work correctly.

Reply:
yes, I know we can, just didn't need to bother for this example, especially as we don't have AP_Int64, but I can add it, negative support is worth it

Reply:
You could get negatives working and still unpack to 32 bit in the script. Maybe return nan if anything other than the sign bit is set in the top 32 bits. That should cover all values that a AP Periph could return.

Reply:
fixed up to handle negative, still limited to 32 bit though. Pity we can't have an int64_t type and string.format() it